utils/docs/pcb_backannotate.1.in
.TH pcb_backannotate 1 "@DATE@" "gEDA Project" @VERSION@

.SH NAME
.B pcb_backannotate
\- backannotates changes from a PCB annotation file to schematic

.SH SYNOPSIS
.B pcb_backannotate \-h\||\-\-help

.B pcb_backannotate \-V\||\-\-version

.B pcb_backannotate
[options]
.I ecofile.eco infile1.sch [infile2.sch ...]


.SH DESCRIPTION

The
.I pcb_backannotate
program reads an engineering change order (ECO) file generated by the PCB program
and backannotates the requested changes to a gschem schematic or schematics.
Refer to the PCB manual for a complete description of the ECO file format.

.SH OPTIONS
.TP
\-h, \-\-help
prints help message and exits
.TP
\-n, \-\-nocopy
Leave the modified schematic files in new files
whose names are generated by appending
.I ".new"
to the
original file names.  The default is to overwrite the original.
.TP
\-v, \-\-verbose
operate verbosely
.TP
\-V, \-\-version
prints version information and exits

.SH EXIT STATUS
.I pcb_backannotate
will return
.B 0
if all files processed successfully,
.B non-zero
if the command did not complete successfully.

.SH BUGS
Rather than parsing and processing schematic files directly, this utility should
be rewritten using the gEDA scheme API.  This would take advantage of the existing
parsers for the schematic files.

.SH SEE ALSO
gnetlist(1), gschem(1), pcb(1)
.PP
.SH COPYRIGHT
.nf
Copyright \(co 2003-@YEAR@ gEDA Contributors.  License GPLv2+: GNU GPL
version 2 or later.  Please see the `COPYING' file included with this
program for full details.
.PP
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


NEWS
==============================================
 'gschem and Friends' Electronic Design Suite
==============================================

Copyright (C) 1998-2015 gEDA Developers

This file documents important user-visible changes in gEDA/gaf.  For
more information, please consult the `ChangeLog' file.

Notable changes in gEDA/gaf 2.1.0.0 - 04/01/15 Rampage Edition

* This release of the Rampage Edition of the gEDA-gaf suite commemorates
  the 17th Anniversary of the first public release of gEDA-gaf. The 2.09
  series releases reflected a very aggressive development of new features
  and major enhancements in many areas. Numerous enhancements focused on
  existing features and functions that did not always work as expected.

* New editing functions provide powerful methods to improve productivity.

  A new Extend Objects functions allows linear objects to be extended to
  other objects with a simple key press.

  A new Break function allows Arcs, Boxes, Circles, Linear objects, and
  Paths, both closed and open, to be broken at user specified points.

  A Powerful new Array dialog has been added to create arrays of selected
  objects. Creating Arrays allows both verb-noun or noun-verb operations.

  A new Auto-Snap Objects to Grid editing function has been added allowing
  Selections to be quickly repositioned to the current Grid snap. The Auto
  Snap features works with all objects types, including graphical and image
  objects.

  A new bimodal Offset function provides editing methods to quickly create
  new objects at programmable distances. The distance can be specified by
  entering a value in a dialog or offsetting Selection sets to the current
  pointer position using hot-keys. The offset distance is retained between
  operations.

  A new Select Last function has been added to select the last object created
  or edited.

  The visibility of netname attributes can now be toggled while inside an
  action.

  A new RC keyword; third-button-cancel, allows user to specify a preference
  for canceling operations using the third mouse button. In addition, object
  adder modes can now be cancel by double-clicking the final entry point.

  A new Zoom to Magnification feature allows user to zoom to a specified level.

* Picture replacement operations can now be undone.

* A new Export functions allow exporting of embedded objects, including image
  objects.

* Picture objects no longer need to maintain the original aspect ratio. Aspect
  ratios of image objects can be freely modified using Grips. The Ctrl key can
  be combined to maintain the aspect ratio identically to the old behavior.
  When exporting images using the new export feature, options are provided to
  export the image at the original or the current aspect ratio and size.

* Coordinates can now be displayed in the status bar using Cartesian and vector
  formats. The format can be changed at any time, include inside of an action,
  by right clicking on the coordinates widget and selecting the format from a
  pop-up menu. Vector formats revert to the previous selected format when not
  applicable to the active context.

* Pointer button assignments can now be set via a pop-up menu by clicking on
  the associated status widget on the Status bar. Assignments can still be set
  using in the Preferences Dialog or over-ridden using RC keywords, but the
  new method allows changing assignments while inside an action.

* The status text now changes to blue when inside of an editing action.

* F-Keys have been assigned to common/standard program functions.

* Menus in Gschem have been revised for better organization with enhanced
  options. The tools menu now includes access to Version tracking and
  alternative simulations back-ends. A toggle widget has been added to
  the Options menu to toggle a new drag-can-move preference variable in
  order to enable and disable dragging moving.

  A new File/Export main menu fly-out was added for options to export both
  embedded components and pictures. Exporting pictures utilized a new custom
  Image Chooser Dialog allowing user to select from a variety of options.
  All available image formats suppported by host libraries are supported.

  Draw-Order functions have been added to the Page menu as a fly-out using
  Scheme, allowing users to manipulate the order in which objects are drawn.

  The pop-up context menu now provides options for Copy, Move, Rotate, Mirror,
  Component Properties, Zoom to Mag, and Zoom to Selection. Additional options
  have also been add as fly-out side menus to the pop-up menu. Fly-out options
  use the X-Y location of the pointer when the context was invoked. Copy, Move,
  Rotate, and Mirror actions were reinstated and work identically to previous
  implementations.

  Additional icons are provided for "Move", Simulations, DRC, and attribute-
  reset. The default icon for "Move" was changed to have a more consistence
  theme with other editing icons.

  Menu tool-tips for File/Recent and Sessions/Restore now respond properly to
  user preferences. Tool-tips for symbol names on the Component dialog field
  shows the full path to symbol file.

  Anti-alias options were added to the Preference dialog in Gschem for the
  anti-alias method to be used by the chosen render. Modifing the anti-alias
  method longer requires a re-start for the new setting to take effect.

* Page Select Dialog has been enhanced to allow displaying file names without
  the paths, a toggle switch allows users to switch the path portion on or off
  in the tree-view. The user perference is retained between sessions.

* The appearance of Auto-Number Dialog has been improved and is more robust.
  The dialog was enhanced with frames and separate filter options for scope
  text including options for wild-cards and a list of recently used filters.
  Recent filters are not retained between sessions.

* The Line Type Dialog now includes an options combo to set the line end-type.


utils/scripts/pcb_backannotate.pl
#!/usr/bin/perl -w
#
# Copyright (C) 2003, 2006, 2010-2012 Dan McMahill
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if  not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA


# This script is used to process annotation files from the PCB
# program (http://pcb.geda-project.org) to backannotate
# changes to gEDA schematics.
#
# It is heavily based on the pads_backannotate program which is also
# part of gEDA

# for parsing input options
use Getopt::Long;

# for ceil function
use POSIX;

# don't allow -he to be interpreted as --help
$Getopt::Long::autoabbrev=0;

# make the options be case sensitive
$Getopt::Long::ignorecase=0;

&GetOptions( ('h|help' => \&usage,
	      'n|nocopy' => \$nocopy,
	      'v|verbose' => \$verbose,
	      'V|version' => \&version
	     ) );

usage() if $Getopt::Long::error;
usage() unless @ARGV;

# Annotation file name
$eco = shift( @ARGV );

# if no schematic names follow, exit
usage() unless @ARGV;

# make sure the input netlist exists and we can open it
$i = 0;
while( @ARGV ) {
  $fname[$i] = shift( @ARGV );
  die "Schematic file $fname[$i] does not exist or can not be read"
    unless -r $fname[$i];
  $i++;
}

$filecnt = $i;


if( $verbose ){ print "Loading PCB annotation file file: $eco\n"; }
my $eco_state="DEFAULT";
open( ECO, "$eco" ) or die "Can't open PCB annotation file $eco !\n";

my $ver = 0;
while( $line = <ECO> ) {
    if( $line =~ /^\*FILEVERSION\* / ) {
	$ver = $line;
	$ver =~ s/^\*FILEVERSION\*\s*//;
	$ver =~ s/[^0-9]*$//;
    } elsif( $line =~ /^\*VERSION\* 20060814/ ) {
	$ver = 20060814;
    }
}
close( ECO );

if( $ver == 0) {
    print "ERROR:  Unable to determine annotation file version.\n";
    exit( 1 );
}

print "Annotation file version is $ver\n" if $verbose ;
if( $ver != 20061031 && $ver != 20060814 ) {
    print "ERROR:  This version of the program is unable to process\n";
    print "        PCB annotation files of version $ver.\n";
    exit( 1 );
}


# reopen and parse the file this time
open( ECO, "$eco" ) or die "Can't open PCB annotation file $eco !\n";
while( $line = <ECO> ) {
    #  if( $verbose ){ print "$line\n"; }
    if( $line =~ /^\*COMMENT\*/ ) {
	print "Skipping comment line: $line\n";
	next;
    } elsif( $line =~ /^\*VERSION\* 20060814/ || $line =~ /^\*FILEVERSION\* [0-9]*/) {
	# the version line.  Note that the very first version of the annotation
	# file used *VERSION* but it was quickly changed to FILEVERSION to avoid
	# confusing it with the program version.
	next;
    } elsif( $line =~ /^\*RENAME\*/ ) {
	# rename refdes in design (forward/backward annotation)
	#$eco_state="RENAME";
	print " *RENAME* (Refdes Renumber)\n" if $verbose;
	$line =~ s/^\*RENAME\*\s*//;
	parseRENAME($line);
	next;
    } elsif( $line =~ /^\*WARN\*/ ) {
	# A warning generated by PCB
	print "Found the following warning in the annotation file:\n";
	print "$line\n";
	next;
    } elsif( $line =~ /^\*/ ) {
	print "WARNING:  Unknown command line:\n";
	print "          $line\n";
	$eco_state="DEFAULT";
	next;
    } else {
	# this must be a data line
	#if( $verbose ){ print "Processing data line: $line"; }
    }

}
close( ECO );

for($i=0; $i < $filecnt; $i++) {
  print "Processing schematic file #", $i+1, ": $fname[$i]\n";
  open(NETLIST,"$fname[$i]") or
    die "Can't open schematic $fname[$i]: $!\n";

  # open output netlist
  $outfname="$fname[$i].new";
  open(OUTSCH,">$outfname") or die "Can't open $outfname: $!\n";

    while($line = <NETLIST>) {
	$line = executeRENAME($line);
	print OUTSCH "$line";
    }
    close(NETLIST);
    close(OUTSCH);

    if( $nocopy ) {
      print "Leaving page #",$i+1," output in $outfname\n";
    }
    else {
      system("mv $outfname $fname[$i]");
    }

}

print "\n---- Gate Swapping ----\n";
executeSWPGATES();
print "\n---- Pin  Swapping ----\n";
executeSWPPINS();
print "\nBackannotation finished ", scalar localtime, "\n";

exit;


#######################################################################
#
# Subroutines
#
#######################################################################

#---------------------------------
# executeRENAME(line)
#---------------------------------

sub executeRENAME {
  my $line = shift(@_);

  return $line unless defined %cmd_rename;
  return $line unless $line =~ /^refdes=/;

  # pick out the reference designator
  $refdes = $line;
  $refdes =~ s/^refdes=//;
  $refdes =~ s/[\r\n]*$//;

  # see if its listed in our hash of reference designators to be
  # renamed
  return $line unless exists $cmd_rename{$refdes};

  print "executeRENAME():  Renaming $refdes to $cmd_rename{$refdes}\n"
      if $verbose;
  return "refdes=$cmd_rename{$refdes}\n";

}

#---------------------------------
# executeSWPGATES()
#---------------------------------

sub executeSWPGATES {
  my $key;

  foreach $key (keys %cmd_swap_gates ) {
    print "Please manually swap gates $key and  $cmd_swap_gates{$key}\n";
  }
}

#---------------------------------
# executeSWPPINS()
#---------------------------------

sub executeSWPPINS {
  my $key;
  my @pins;

  foreach  $key (keys %cmd_swap_pins ) {
    @pins = split '\.',,$cmd_swap_pins{$key};
    print "Please manually swap pins $pins[0] and $pins[1] on $key\n";
  }
}

#---------------------------------
# parseRENAME(line)
#---------------------------------

sub parseRENAME {
  my $line = shift(@_);
  my @refs;
  @refs = split ' ',,$line;

  $refs[0] =~ s/"//g; # "
  $refs[1] =~ s/"//g; # "

  print "parseRENAME():  Scheduling rename of $refs[0] to $refs[1]\n"
      if $verbose;
  $cmd_rename{$refs[0]} = $refs[1];

}

#---------------------------------
# parseSWPGATES(line)
#---------------------------------

sub parseSWPGATES {
  my $line = shift(@_);
  my @refs;
  @refs = split ' ',,$line;

  print "parseSWPGATES():  Scheduling swap of gate $refs[0] with $refs[1]\n"
      if $verbose;
  $cmd_swap_gates{$refs[0]} = $refs[1];

}

#---------------------------------
# parseSWPPINS(line)
#---------------------------------

sub parseSWPPINS {
  my $line = shift(@_);
  @refs = split ' ',,$line;
  @pins = split '\.',,$refs[1];

  print "parseSWPPINS():  Scheduling swap of pins on $refs[0] : pins ",
  "$pins[0] and $pins[1]\n" if $verbose;
  $cmd_swap_pins{$refs[0]} = $refs[1];

}

#---------------------------------
# usage()
#
# prints program usage
#---------------------------------

sub usage {
  my $pname = $0;
  $pname =~ s/.*\///g;

  print <<EOF;

Usage:

  $pname [-n|--nocopy] [-v|--verbose] change.log file1.sch [file2.sch ...]
  $pname -h|--help\n
  $pname -V|--version\n

$pname reads a PCB annotation file and backannotates changes\n
to a gEDA schematic.

$pname accepts the following options:

  -h
  --help      Displays this help message.

  -n
  --nocopy    If given, this flag leaves the modified files in new files
              whose names are generated by appending a \".new\" to the
              original file names.  The default is to overwrite the original.

  -v
  --verbose   Enables verbose output.

  -V
  --version   Shows the version of this program.


$pname was written by Dan McMahill <dmcmahill\@netbsd.org>


EOF

  exit;
}

#---------------------------------
# version()
#
# prints program version
#---------------------------------

sub version {
  open(PROG,"$0") or die "Could not open \"$0\" to find version\n\n";
  my $pname = $0;
  $pname =~ s/.*\///g;

  while($line = <PROG>) {
    if( $line =~ /^#\s*\$Id.*\$/) {
      @words = split ' ',,$line;
      $version = $words[3];
      $date = $words[4];
      print "$pname ($0):  Version $version, $date\n";
      exit;
    }
  }
  print "Could not determine version of \"$0\"\n\n";
  exit;
}

